### PR TITLE
CI: Add clang to GitHub Actions build-tests matrix

### PR DESCRIPTION
**Pull request by kinkie**
Add "compiler" dimension to build-tests matrix.

Do not test layer-04-noauth-everything to partially compensate for the
new matrix dimension repeating other layer tests 3 times.

Change build-tests artifacts name pattern to include compiler dimension
to avoid name clashes that would result in logs being overwritten and
lost. Also use target environment instead of runner environment in
anticipation of adding container-based tests where the two differ.

### CODE REVIEW

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for this CI improvement!

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:104]: We don't even try gcc on macos. The target on that platform is Xcode, aka clang. Adding a comment

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:117]: But we may also drop this requirement to install everything every time
and use the pre-build docker images we maintain.
This would free us from the constraint of only using the couple of
Ubuntu versions that github supports, and give us a full alternative to
jenkins

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml]: > Please try to refactor to avoid as many of the following red flags as possible:
> 
> * Needlessly different commands for different OSes. I am aware that MacOS tests currently exclude gcc; ignore that ephemeral fact when refactoring.

There is little point in keeping them aligned, but sure.

> * CXX is only set for some compilers. I am aware that our ./configure does not need CXX to be set when CC is gcc; ignore that compiler-specific fact when refactoring.

It's not "our" configure. It's baked into any configure script.

> * An inconsistent mixture of `env FOO` and `export BAR` commands.

It's unavoidable, because "env:" commands do not support string manipulation. The next best option is to do it all in the shell script. Doesn't feel like an improvement to me, but ¯\_(ツ)_/¯

> * Avoidable conditionals

There is one unavoidable conditional


> One of the tools that may help you in this refactoring is this matrix trick:
> 
> ```yaml
> matrix:
>     os: [ ubuntu-22.04, macos-14 ]
>     compiler: [ {cxx: g++, cc: gcc}, {cxx: clang++, cc: clang} ]
> ```
> 
> or, if you prefer, this matrix trick:
> 
> ```yaml
> matrix:
>     os:
>         - ubuntu-22.04
>         - macos-14
> 
>     include:
>        - cxx_compiler: g++
>          c_compiler: gcc,
>        - cxx_compiler: clang++
>          c_compiler: clang,
> ```
> 
> I have neither [invented](https://stackoverflow.com/q/66025220) nor tested these tricks, but I do hope there is something better (like these tricks) than pairing compiler names using shell `test` statements!

We could use shell "if" statements and "case" statements as well :P

Interestingly, github's code editor complains about using a map for a matrix dimension (although it works). This could mean it might stop working in the future. Using it uglifies the exclusion list, but [it works](https://github.com/squid-cache/squid/actions/runs/9037244905/job/24835847230).

I'm following your suggestion, although unconvinced. Hopefully it's code we won't need to touch ever again

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:162]: Using runner.os here is just logically wrong.
runner.os is Linux for ubuntu-22.04, ubuntu-20.04, freebsd (which runs on ubuntu-latest), openbsd (ditto) and who knows what else. We need to add matrix.compiler here, otherwise logs for clang and gcc will clash and overwrite each other. Mix-n-match of runner.os and matrix.compiler makes no sense at all.

The alternative is to run a PR to change "runner" to "matrix" standalone, and wait for things to settle and deal with merge conflicts.

Comment (kinkie) [reply .github/workflows/default.yaml:162]: I had initially forgotten to add it, and noticed that artefacts overwrote each other depending on timing.

Comment (kinkie) [reply .github/workflows/default.yaml:162]: In other words: if we want to progressively replace jenkins with github workflows, we can't avoid this change. If it belongs to this PR or to another PR, is purely a matter of convenience

Comment (rousskov) [reply .github/workflows/default.yaml:161]: >> If there is a good reason for doing this change here and now, please disclose that reason in PR description.

> We need to add matrix.compiler here, otherwise logs for clang and gcc will clash and overwrite each other. 

Please add that info to the PR description. Adding layers to the matrix in #1787 did not result in clashes [AFAICT](https://github.com/squid-cache/squid/pull/1787/files#r1571154576), but I was not sure whether GitHub is smart enough to handle clashes when adding compilers. (I am aware of the difference in log file names between those two use cases.) Thank you for answering my question.

Comment (kinkie) [reply .github/workflows/default.yaml:161]: > but I was not sure whether GitHub is smart enough to handle clashes when adding compilers

The key difference between these two cases is that in this cases, archived filenames are the same so there is no way to collate them.

File: .github/workflows/default.yaml
Comment (rousskov) [reply .github/workflows/default.yaml:118]: > But we may also drop this requirement to install everything every time and use the pre-build docker images we maintain. 

The current approach is much better for the existing simple cases IMO. I can detail why, but this PR is probably not the best place to debate what future unrelated changes are beneficial.

> This would free us from the constraint of only using the couple of Ubuntu versions that github supports, and give us a full alternative to jenkins

Please do not interpret my response above as a vote against using docker images on GitHub Actions. The two approaches are not mutually exclusive (although they may benefit from being located in different Action _jobs_).

File: .github/workflows/default.yaml
Comment (rousskov) [reply .github/workflows/default.yaml:161]: > > I am aware of the difference in log file names between those two use cases.

> The key difference between these two cases is that in this cases, archived filenames are the same so there is no way to collate them.

As I said, I am aware of this file naming difference between the two cases. I suspect there is a way to avoid multiplying the number of top-level artifacts, but, at this time, I am not suggesting that we spend time on that.

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:104]: The above comment (pretty much retells what the code already says but) does not explain _why_ we are not testing gcc on MacOS. Please rephrase. For example, if you just did not have time or desire to investigate what gcc-specific workarounds are needed on MacOS, then we can do something like this:

```suggestion
            # TODO: Investigate what it would take to build Squid with gcc on MacOS
```

BTW, [AFAICT](https://en.wikipedia.org/wiki/Xcode), Xcode is an IDE. We are not using IDEs to test Squid builds here. AFAICT, gcc is _not_ incompatible with Xcode (even though it is no longer included with Xcode). Thus, referring to Xcode does not seem sufficient (at least).

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:137]: If adding support for clang tests does not require adding this option, let's not add it in this PR:

```suggestion
          ./test-builds.sh
```

Otherwise, please add a YAML comment to explain why removing build logs is a good idea when building with clang.

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:135]: Is it too dangerous to make this a job-specific environment (rather than step-specific environment) because some earlier steps may pick up a not-yet-installed compiler?

If it is too dangerous, we would need to duplicate this for MacOS. The latter is best done now (rather than when we add another MacOS compiler variation) IMO but I do not insist on that.

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:104]: AFAIK Xcode is not just an IDE. It contains the compiler, and more important all the APIs, headers, and bits needed to target the various Apple operating systems.

gcc is available on homebrew, but I don't think it's worth targeting because:
- it's highly unlikely anyone wanting to build Squid on MacOS would use gcc to build squid. It might even be that gcc relies Xcode for system includes, frameworks, and libraries
- there is little signal to be obtained in doing it - we are already testing dual-compiler platforms to avoid dialects slipping in and locking us into a specific compiler
- I tried a build with gcc-13, it failed on not finding any needed symbol in libcppunit. While I have not investigated in dhetail, a fitting explanation might be that gcc does not fully conform to the clang ABI. This would send us in library ABI version hell

Comment (rousskov) [reply .github/workflows/default.yaml:109]: If want this PR to rule out testing with non-clang compilers on MacOS, then we can do something like this instead of the earlier suggestion:

```suggestion
            # Non-clang testing on MacOS is too much work for very little gain.
```

Comment (kinkie) [reply .github/workflows/default.yaml:109]: > If you want this PR to rule out testing with non-clang compilers on MacOS, then we can do something like this instead of the earlier suggestion

This wording sounds good to me. Of course I don't rule out anything, I am just unwilling to put in the effort myself as I don't think it's worth it

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:111]: Done; chose an even more compact version using slashes to separate dimensions

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml]: [Seems to be working](https://github.com/squid-cache/squid/actions/runs/9049941451). It might come back and haunt us in the future (e.g. if we need to specify a path to run a build in some bizarre environment) but for now it seems to be working

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:127]: Let us leave installing a full environment on, for now. We will be able to revise in the future if there is benefit

Comment (rousskov) [reply .github/workflows/default.yaml:127]: > Let us leave installing a full environment on, for now. We will be able to revise in the future if there is benefit

Your suggestion is compatible with this change request IMO, but we may define "full environment" differently, or perhaps the discussion drifted away too far from the change request itself.

If it is possible to address this change request without heroic efforts,  please do that (to provide the benefits this change request outlines). I see no signs that it is impossible.

FWIW, I would lean towards the second sketch among the two suggested earlier:

```suggestion
          sudo apt-get --quiet=2 install linuxdoc-tools libtool-bin ${{ matrix.compiler.CC }}
```

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:143]: Copying from a seemingly forgotten (and now outdated) [change request](https://github.com/squid-cache/squid/pull/1812#discussion_r1597260377):

```suggestion
        run: ./test-builds.sh ${{ matrix.layer.layer }}
```

.... or, if the layer.layer change request is addressed:

```suggestion
        run: ./test-builds.sh ${{ matrix.layer.name }}
```

```suggestion
        run: ./test-builds.sh ${{ matrix.layer.test-builds-parameter }}
```

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:107]: Please avoid visually odd layer.layer names. I can suggest using "name" for the inner "layer" field:

```suggestion
          - { name: layer-00-default, nick: default }
          - { name: layer-01-minimal, nick: minimal }
          - { name: layer-02-maximus, nick: maximus }
          - { name: layer-04-noauth-everything, nick: noauth }
```

Another reasonable option is "test-builds-parameter" (because that is why we need layer.layer value; it is how it is used).

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:199]:
```suggestion
```

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:113]: I would separate these three configuration sections a bit, especially since one of them is multiline.

```suggestion

    name: build-tests(${{ matrix.os }}/${{ matrix.compiler.CC }}/${{ matrix.layer.nick }})

```

I also think a comma is an overall better separator than a slash for this human-read info, but I do not insist on using a comma:

```suggestion

    name: build-tests(${{ matrix.os }},${{ matrix.compiler.CC }},${{ matrix.layer.nick }})

```

**Review by rousskov — CHANGES_REQUESTED** [review]
Great progress, thank you.

File: .github/workflows/default.yaml
Comment (kinkie) [reply .github/workflows/default.yaml:127]: Are you saying you will block this PR unless this change is implemented?
I'm not keen on it, because it makes the build environments more different than strictly needed, so unless it's the only way to land the change, I would prefer to not do it.

Comment (rousskov) [reply .github/workflows/default.yaml:129]: > Are you saying you will block this PR unless this change is implemented?

No, I am not saying that. I am also not saying that this PR will be approved without changes. As you know, I strive to judge code based on its merits rather than use ultimatums and power plays. If you find errors in my logic, and the fixed logic leads to a different decision, then either no or different PR changes will be needed.

> I'm not keen on it, because it makes the build environments more different than strictly needed

Thank you for disclosing the basis for your proposal to install clang where it is not used. IMO, the priority/weight of that "similar build environments" goal is far lower/smaller than that of "more representative build environments" and "more pure build environments" goals[^G]. The "similarity" goal has little to do with test realism/quality/usefulness -- similarity is primarily a test code clarity and maintenance simplification goal. The suggested change does not complicate test code understanding and maintenance by much (if at all).

[^G]: These short "quoted" summaries may not reflect all the important nuances of each goal characteristic well, but I hope you get the idea.

In the ideal world of unlimited resources, we would have tested all four combinations:

1. Using gcc on Ubuntu with gcc and without clang installed.
2. Using gcc on Ubuntu with gcc and with clang installed.
3. Using clang on Ubuntu without gcc and with clang installed.
4. Using clang on Ubuntu with gcc and with clang installed.

This PR proposes to pick two out of four possible builds. I assume[^A] that removing (or not installing) gcc is difficult, so we cross out the only "without gcc" item -- item 3. After that, we have to use item 4 to get any clang tests. Out of the remaining two builds (item 1 and item 2), the first one is arguably more "representative" (fewer folks install _both_ compilers) and clearly more "pure" (i.e. has fewer unused components) than the second. Thus, item 2 looses, and we are left with the final two:

1: Using gcc on Ubuntu with gcc and without clang installed.
4: Using clang on Ubuntu with gcc and with clang installed.

[^A]: As always, please correct any wrong assumptions. If this assumption is wrong, then we should probably use "pure" items 1 and 3 instead of items 1 and 4.

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:162]:
```suggestion
          ./test-builds.sh ${{ matrix.layer.name }}
```

BTW, these leftovers exposed a small test-builds.sh bug: That script mishandles an empty string passed as an argument. The current PR code does _not_ pass an empty string to test-builds.sh[^E]. I was going to suggest that we modify this PR line, so that it does pass an empty string if we make a similar mistake in spelling matrix.layer.name, but, after testing, I am _not_ suggesting that because of that test-builds.sh bug that has no trivial fix (that I could find).

[^E]: Instead, test-builds.sh gets no arguments at all (and correctly tests all layers) when called from this current PR code line.

File: .github/workflows/default.yaml
Comment (rousskov) [inline .github/workflows/default.yaml:107]: We have discussed removing this layer from these tests. If you want to remove this layer in _this_ PR, please commit the following suggestion:

```suggestion
```

One could argue that removing this layer is in this PR scope because this PR adds more tests (by adding a compiler dimension to the matric), and we are worried about having too many tests (in total).

**Review by rousskov — APPROVED** [review]
I am approving after merging earlier suggestions as discussed during the meeting and polishing PR description a bit. We will need to change the set of the required tests and Anubis configuration to land this PR, but I am requesting @yadij review first.

**Comment by rousskov** [comment]
I also updated PR description to explain why we now use target environment instead of runner environment in build-tests artifacts name pattern.

**Review by yadij — APPROVED** [review]

**Review by rousskov — APPROVED** [review]

**Comment by rousskov** [comment]
I have explained layer-04-noauth-everything removal in the PR description.

**Comment by rousskov** [comment]
I have completed items 2 and 3 on the test upgrade [plan](https://github.com/squid-cache/squid/pull/1785#issuecomment-2066676266). This allowed Anubis to stage this PR. @eduard-bagdasaryan, please update Anubis configuration to reflect the current set of required status checks. Doing so will complete item 4.

**Comment by eduard-bagdasaryan** [comment]
I updated that number as the number of tests in [5a7811a](https://github.com/squid-cache/squid/commit/5a7811aff6919406fa93317579c510c6b6708aec). It is 25 now : 12 GitHub tests + 12 Jenkins tests + PR approval.

**Comment by rousskov** [comment]
@eduard-bagdasaryan, thank you. This update leaves only one last step on that upgrade [plan](https://github.com/squid-cache/squid/pull/1785#issuecomment-2066676266):

6: Merge master into other pending PRs as needed. They will not be auto-merged until their branches get new tests (because GitHub testing requirements have changed).

There is no reason to rush and update all PRs, of course. We should do that "as needed", usually around the time when a PR is cleared for merging.

**Comment by eduard-bagdasaryan** [comment]
> It is 25 now

It looks like that one of GitHub tests in [5a7811a](https://github.com/squid-cache/squid/commit/5a7811aff6919406fa93317579c510c6b6708aec) is not required: "GitHub CI / CodeQL-tests". I'll fix that parameter to 24.

**Comment by rousskov** [comment]
> It looks like that one of GitHub tests in [5a7811a](https://github.com/squid-cache/squid/commit/5a7811aff6919406fa93317579c510c6b6708aec) is not required: "GitHub CI / CodeQL-tests".

@yadij, @kinkie, any objections to making that test required for master PRs? FWIW, I do not recall seeing any false negatives from that test.

**Comment by kinkie** [comment]
No objection

**Pull request closed without merging**
